add husky to run tests, format and lint code on commit

lukeecart opened the pull request:
### Description

Suggestion to run husky on commit to run tests, format and lint code. 
The only setup for future contributors is run `yarn install` instead of creating a git hook.
closes #264 

### Type of Change

<!--- Please check the boxes that are relevant to this PR: -->

- [X] Adding a new entry
- [ ] Editing an existing entry (fixing a typo, bug, issues, etc)
- [ ] Updating the documentation

### Checklist

<!--- Please check the boxes that are relevant to this PR: -->

- [X] My entry follows the Codecademy Docs style guide
- [X] I have performed a self-review of my own writing and code
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have checked my entry and corrected any misspellings
- [X] I have confirmed my changes are not being pushed from my forked `main` branch
- [X] I have confirmed that I'm pushing from a new branch named after the changes I'm making
- [X] All writings are my own

<!---
Having trouble with the PR checker? Here are some common issues and resolutions:

- verify_formatting is failing
  - run `yarn format path/to/markdown/file.md` or `yarn format:all` and commit the results
- verify_lint is failing
  - same as above
  - if verify_lint is still failing, running `yarn lint` locally should let you know what needs to be changed by hand
- test is failing
  - ensure any new markdown files have a `Title` and `Description` defined in their metadata
  - ensure any new markdown files only contain alphanumerics and dashes in their file names and have the same name as their parent directory
  - if that looks ok, running `yarn test` locally should let you know what the issue is

If you want your Codecademy username to be displayed in your entry (feature coming in October), make sure to have your Codecademy user profile linked to your GitHub:

1. Go to your Codecademy dashboard.
2. Click on your profile image in the top-right and then choose "Profile".
3. Then click "Edit Profile".
4. In the GitHub field, add your GitHub profile URL.
-->

Dusch4593 commented:
@christian-dinh @lukeecart FWIW, where do you think we stand with this PR given [we recently switched from CircleCI to GitHub Actions](https://github.com/Codecademy/docs/pull/392) for running tests that include lint/format?

christian-dinh commented:
@Dusch4593 I think this is still valuable! The point of this change is to catch lint/format issues before they hit the CI in the first place, so whether we're using circleci or gh actions shouldn't affect it!

christian-dinh commented on .husky/pre-commit:
```diff
@@ -0,0 +1,7 @@
+
+npm run lint
+npm run format:all
+git add .
```
no worries at all, sorry I didn't respond sooner. I'm also not 100% on hooks 😅 

if you're cool with it I can take a look at some of our lint-staged configs and see if I can get them to work on this branch sometime next week

lukeecart replied:
@christian-dinh  yeah sounds good. I'm conscious that this PR is getting a bit stale but I still think there is value in adding hooks

Dusch4593 commented:
> @Dusch4593 I think this is still valuable! The point of this change is to catch lint/format issues before they hit the CI in the first place, so whether we're using circleci or gh actions shouldn't affect it!

👀 Gotcha! You're right, this will make things a lot easier for everyone! Will this also mean fewer commits to ultimately squash before merging?

lukeecart commented:
@Dusch4593 @christian-dinh Hey guys. I've finally got around to working on this one again.
I've added lint-staged as suggested which runs a pre commit hook to lint and format all the files that have been staged.

This is ran when a contributor commits their files. This will format and lint their files before it gets to the github actions CI :D This should reduce the number of commits that are pushed to address CI failures 🎉.
Sorry this PR has been a bit stale...

Dusch4593 commented:
@lukeecart, thank you for your work on this! I actually really enjoyed tuning into this PR. Got to learn some nifty workflow concepts and tools. This change is _much_ appreciated. 

cc: @christian-dinh

Wanted to check in with you and Luke on this:

> This is ran when a contributor commits their files. This will format and lint their files before it gets to the github actions CI 

Doesn't `lint-staged` execute when changes are added, or _staged_, for commit since it's being used in the pre-commit hook?

lukeecart commented:
> Wanted to check in with you and Luke on this:
> 
> > This is ran when a contributor commits their files. This will format and lint their files before it gets to the github actions CI
> 
> Doesn't `lint-staged` execute when changes are added, or _staged_, for commit since it's being used in the pre-commit hook?

Yeah the listing and formatting is ran on the staged files (I.e. git add ...) when git commit it ran

lukeecart commented:
@Dusch4593 @christian-dinh  Is there anything else required to merge this PR?
Have you both been able to test it out?

Dusch4593 commented:
> @Dusch4593 @christian-dinh 👋 Is there anything else required to merge this PR?
Have you both been able to test it out?

I'm testing right now by:

1. Making edits to two topic files (`python.md` and `markdown.md`)
2. Adding/committing one of them.

While testing the hook, I ran into a couple things. Here are my thoughts and observations:

- The `.husky/pre-commit` hook still isn't set as executable. I recommend running `chmod +x .husky/pre-commit` and pushing to this PR.
- The way the commands are written on line 37 in `package.json` makes Prettier think that `&& npm run lint` are 4 separate files that need formatted. Since these should be treated as two separate scripts, change the value of the `*.md` property in `lint-staged` to an array of those commands:

```
lint-staged:
  "*.md": [
    "npm run format",
    "npm run lint"
  ]
```

- After making that change and retrying things by staging changes in `content/python/python.md`, the `format` seemed to work but when running `npm run lint`, the linter catches the unstaged `markdown.md` file and I get the following: 

<img width="619" alt="Screen Shot 2021-12-26 at 9 30 04 AM" src="https://user-images.githubusercontent.com/23144845/147411186-beaa3528-52d7-4878-855f-4e30a449ace2.png">

The `lint` script implicitly runs `lint:md` and `lint:js`. In `lint:md`, it seems to be checking every `.md` file in every folder? 

```
"lint:md": "markdownlint \"./content/**/*.md\""
```

That's as far as I got with things for now. I'll keep tinkering on my fork but I feel we are definitely on the way to making this a reality! 😃

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Codecademy/docs?pullRequest=265) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Codecademy/docs?pullRequest=265) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: lukeecart<br/>:x: lukeca362<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Codecademy/docs?pullRequest=265) it.</sub>

lukeecart commented:
@Dusch4593 I've rebased, and merged commits to chmod the pre-commit file and also added a new command in package.json to format just md files

Dusch4593 commented:
> @Dusch4593 I've rebased, and merged commits to chmod the pre-commit file and also added a new command in package.json to format just md files

Nice thanks Luke! I'll test this out later this morning. 😄

Dusch4593 commented:
@lukeecart I just tested out your changes and we are definitely on the right track, I think! When I run `add`/`commit`, though, both my staged file and my unstaged file are formatted. This is because the `format:md` script encapsulates _every_ `.md` file in _every_ folder. The same goes for the `lint:md` script. 

If I change `format:md` to just `format` and test again, the `format`-part of the hook passes but not the `lint:md`-part since the unstaged file's error was detected.

```
  "lint-staged": {
    "*.md": [
      "npm run format",
      "npm run lint:md"
    ]
  }
```

<img width="619" alt="Screen Shot 2021-12-28 at 12 41 34 PM" src="https://user-images.githubusercontent.com/23144845/147592676-e38ae3a5-4e63-4d92-8f88-d05c7863e16f.png">

If I understand the original premise of this PR, we only want the staged files to be formatted/linted prior to commit. So in my test, I want `python.md` to be affected by husky/lint-staged but not `markdown.md`. I went ahead and made the following change to make this happen.

- In `pre-commit`, I added an `--allow-empty` commits flag since I was running into this problem frequently.

```
npx lint-staged --allow-empty
```

- In `package.json`, in the `lint-staged` section, I changed the `*.md` field to have this:

```
  "lint-staged": {
    "*.md": [
      "npm run format",
      "markdownlint"
    ]
  }
```

Running just `markdownlint` will lint any staged `*.md` file detected by `lint-staged`.

<img width="624" alt="Screen Shot 2021-12-28 at 12 45 43 PM" src="https://user-images.githubusercontent.com/23144845/147592997-aea1f8f8-1726-431d-8131-7fce6fbc39db.png">

lukeecart commented:
Sounds good. I think the markdownlint would be a better option. It's a tricky one to get right but once we do this will hopefully reduce the number of commits in PRs trying to resolve formatting

Dusch4593 commented on .husky/pre-commit:
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+. "$(dirname "$0")/_/husky.sh"
+
+npx lint-staged
```
```suggestion
npx lint-staged --allow-empty
```

Dusch4593 commented on package.json:
```diff
@@ -21,13 +23,21 @@
+  "version": "1.0.0",
+  "lint-staged": {
+    "*.md": [
+      "npm run format:md",
+      "npm run lint:md"
```
```suggestion
      "npm run format",
      "markdownlint"
```

Dusch4593 commented:
@lukeecart do tests run as expected on your machine?

lukeecart commented:
@Dusch4593 I'm unable to npm install and get an error with - ERESOLVE unable to resolve dependency tree.


<img width="573" alt="Screenshot 2021-12-29 at 16 13 59" src="https://user-images.githubusercontent.com/22771812/147682098-b5263674-eadd-4538-8a48-6453f8faf589.png">

lukeecart commented:
@Dusch4593 When running yarn install I get errors of a @codecademy/eslint-config incorrect peer dependencies. should this be resolved in this package. I think that package is private.

Once I yarn install, there is an update to the yarn.lock and npm run test passes ✅

Dusch4593 commented:
> Once I yarn install, there is an update to the yarn.lock and npm run test passes ✅

@lukeecart So you're saying that this last time you ran `yarn install` the tests ran fine?

lukeecart commented:
@Dusch4593 Yeah, I literally just ran `yarn install`  then `npm run test`. 

When I was installing it notes there are incorrect peer dependencies but still installs

Dusch4593 commented:
> @Dusch4593 Yeah, I literally just ran `yarn install` then `npm run test`.
> 
> When I was installing it notes there are incorrect peer dependencies but still installs

Yeah, peer dependencies are kind of a pain, sometimes. According to my research, this warning is just giving a heads up about certain packages that need updating within the dependency tree. It's not incorrect, though. This is the breakdown of available updates I'm getting:

<img width="368" alt="Screen Shot 2021-12-29 at 11 34 32 AM" src="https://user-images.githubusercontent.com/23144845/147684213-1a7f6b94-30d2-45e4-abc8-c0821d2301fb.png">

I would defer to @christian-dinh on this one since I don't have a hand in those internal tools.

Dusch4593 commented:
@lukeecart But what I meant by "tests" was whether `lint-staged` does what it is expected to do. You test with the following steps:

1. Make sure `yarn.lock` file is installed (you seem to have done this already)
2. Remove the trailing-empty-line at the end of the `python.md` and `markdown.md` files in `content/python` and `content/markdown`, respectively.
3. Run `git status` to confirm that both files are changed but _un_staged.
4. Run `git add content/python/python.md` to stage only the `python.md` file.
5. Re-run `git status` to confirm `python.md` is staged and `markdown.md` is not.
6. Run `git commit -m "testing husky hook"` to trigger `lint-staged` tasks.
7. Only the `python.md` file should be linted/formatted and the `markdown.md` file remains untouched.

lukeecart commented:
@Dusch4593  Yeah, I just followed those steps on my macbook and a commit has been created with the markdown file unchanged. 
Because of the formatting the commit is basically empty so we should remove `--allow-empty`. 

I also added more text to the python.md file and removed the extra line and that commit also worked.

Dusch4593 commented:
> Because of the formatting the commit is basically empty so we should remove --allow-empty.

@lukeecart No it should be there because otherwise `lint-stage` will abort the entire commit due to presence of an empty commit.

JoshuaKGoldberg commented:
> internal tools

FWIW all the `@codecademy/*` packages mentioned here should exist in github.com/Codecademy/client-modules or github.com/Codecademy/gamut ✨

JoshuaKGoldberg commented on package.json:
```diff
@@ -21,13 +23,21 @@
+      "npm run format",
+      "markdownlint"
+    ]
+  }
```
Nittiest of nits: up for moving the `package.json` keys towards being alphabetically sorted by moving this up? 😄

Dusch4593 replied:
@JoshuaKGoldberg how do things look now? e7a1ee0320b4a5103c5793c6e2484c918557bf2f

JoshuaKGoldberg replied:
🎉

JoshuaKGoldberg reviewed (approved):
Played with it locally and it works great, thanks! 🙌

lukeecart commented:
Thank you @JoshuaKGoldberg and @Dusch4593 for the reviews. Hopefully this is a helpful new feature to assist with new PRs

Dusch4593 commented:
How do things look for this PR from your end, @christian-dinh?

christian-dinh reviewed (approved):
Sorry I was MIA for most of this, but tested locally and everything looks great! Thank you all for getting this through, I think this will help ease a lot of little pains people are having ✨
